### PR TITLE
support for simple kinematic motion

### PR DESCRIPTION
**Pull request by AndrewMeadows**
If you give non-zero velocity/angularVelocity/gravity to an otherwise static entity it will adopt kinematic motion and begin inexorable motion.  Damping and gravity are supported.

### CODE REVIEW

File: libraries/physics/src/EntityMotionState.cpp [hunk 1]
```diff
@@ -40,13 +42,24 @@ EntityMotionState::~EntityMotionState() {
...
 MotionType EntityMotionState::computeMotionType() const {
-    // HACK: According to EntityTree the meaning of "static" is "not moving" whereas
-    // to Bullet it means "can't move".  For demo purposes we temporarily interpret
-    // Entity::weightless to mean Bullet::static.
-    return _entity->getCollisionsWillMove() ? MOTION_TYPE_DYNAMIC : MOTION_TYPE_STATIC;
+    if (_entity->getCollisionsWillMove()) {
+        return MOTION_TYPE_DYNAMIC;
+    }
+    return _entity->isMoving() ?  MOTION_TYPE_KINEMATIC : MOTION_TYPE_STATIC;
+}
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.cpp]: Entities now suport all three motion types:  DYNAMIC, KINEMATIC, and STATIC.

File: libraries/physics/src/EntityMotionState.cpp [hunk 2]
```diff
@@ -215,3 +231,16 @@ void EntityMotionState::sendUpdate(OctreeEditPacketSender* packetSender, uint32_
...
+    
+    // we add DIRTY_MOTION_TYPE if the body's motion type disagrees with entity velocity settings
+    int bodyFlags = _body->getCollisionFlags();
+    bool isMoving = _entity->isMoving();
+    if (((bodyFlags & btCollisionObject::CF_STATIC_OBJECT) && isMoving) ||
+            (bodyFlags & btCollisionObject::CF_KINEMATIC_OBJECT && !isMoving)) {
+        dirtyFlags |= EntityItem::DIRTY_MOTION_TYPE; 
+    }
+    return dirtyFlags;
+}
```
Comment (AndrewMeadows) [inline libraries/physics/src/EntityMotionState.cpp]: Changing velocity is normally an "easy" update for the physics engine (does not require removal/insertion of the body) however changing the motion type is a "hard" update.  We watch for incompatible motion types here and set DIRTY_MOTION_TYPE here (thereby making it a "hard" update).

File: libraries/physics/src/ObjectMotionState.cpp [hunk 3]
```diff
@@ -110,30 +116,23 @@ bool ObjectMotionState::doesNotNeedToSendUpdate() const {
...
-            return true;
+        } else {
+            const float NON_MOVING_UPDATE_PERIOD = 1.0f;
+            if (dt > NON_MOVING_UPDATE_PERIOD && _numNonMovingUpdates < MAX_NUM_NON_MOVING_UPDATES) {
+                // RELIABLE_SEND_HACK: since we're not yet using a reliable method for non-moving update packets we repeat these
+                // at a faster rate than the MAX period above, and only send a limited number of them.
+                return true;
+            }
         }
     }
```
Comment (AndrewMeadows) [inline libraries/physics/src/ObjectMotionState.cpp]: This work removes the "pessimistic" send of entity updates even when when the server-side prediction has not been violated.  Someday when we have a better understanding of how packet-loss affects EntityServer <--> Client communication we will revisit this spot.

**Comment by ZappoMan** [comment]
test this please

**Comment by PhilipRosedale** [comment]
Andrew:  When I create a new entity in JS, how do I distinguish between kinematic and physical/colliding?  Is it the collisionsWillMove flag?

**Comment by AndrewMeadows** [comment]
There are three kinds of entity motion: DYNAMIC, KINEMATIC, and STATIC.

This is how they are categorized:

If collisionsWillMove is TRUE it is DYNAMIC.

Else, if the entity has non-zero linear or angular velocity it is KINEMATIC(*)(**)

Else it is STATIC (***)

(*) KINEMATIC motion supports damping and gravity, so user beware!  The UI of the editEntities.js script gives plenty of rope for users to hang themselves because it applies any small change immediately.

(**) A damped KINEMATIC object that slows to a stop by damping won't go STATIC... it will be KINEMATIC without movement -- which means it will be put on Bullet's "inactive" list until some other object moves nearby to wake it up at which point it will try to compute zero kinematic motion which results in a few wasted CPU cycles.

(***) A STATIC object that gets non-zero velocity (via properties edit) will automatically become KINEMATIC and will remain so until some property changes (via properties edit, not by damping simulation) at which point its motion type is re-evaluated and properly updated.

BTW, the failed build is because Bullet is not being found on Windows and Ubuntu.  This might be related to changes being made for the "require Bullet" PR.  I'll try to coordinate with Leo and continue to ask for assistance from Birarda on making a custom FindBullet.cmak module.

**Pull request merged**